### DISCUSSION TITLE
Is autoremove broken?

### DISCUSSION THREAD

**Discussion by kruegercharles** [Homebrew/discussions#3787] [Everyday usage] [answered]
When I run `brew autoremove` it says:

> ==> Autoremoving 1 unneeded formula:
jpeg
Error: Refusing to uninstall /opt/homebrew/Cellar/jpeg/9e
because it is required by openjpeg, which is currently installed.
You can override this and force removal with:
  brew uninstall --ignore-dependencies 

My understanding is: `autoremove` removes dependencies which are currently _not_ needed. 
Then why does it show me a dependency, which IS currently needed? If it is needed, then I obviously do not want to remove it with `autoremove`. The whole point of `autoremove` is getting rid of _unnecessary_ dependencies. 
Am I missing something?

Also, `autoremove` has no flag for ignoring dependencies. At least they are not listed at `brew help autoremove`. So I cannot ignore this dependency which causes `autoremove` to fail and thereby not removing _any_ unnecessary dependencies. Because of this I recently had to remove a bunch of old dependencies by hand which was not great. 

Is there a way to fix this?

Thank you!

**Comment by gromgit** [comment, marked as answer]
It's almost certainly related to https://github.com/Homebrew/discussions/discussions/3704, i.e. a stale dependency record.

Specifically, `openjpeg` installed before Jul 25 would've pulled in `jpeg` as an indirect dependency via `libtiff`. `libtiff` switched to `jpeg-turbo` at that point, but all Homebrew formulae record their full dependency list _at install time_, hence the discrepancy.

Simply `brew reinstall openjpeg`, then `brew autoremove` should correctly remove `jpeg`:
```bash
$ brew deps openjpeg
jpeg
libpng
libtiff
little-cms2
zlib

$ brew reinstall openjpeg
==> Downloading https://ghcr.io/v2/homebrew/core/openjpeg/manifests/2.5.0-1
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/openjpeg/blobs/sha256:9e57a66212ac4fca6dd1c802c80cea7e3157157b16577d549f2b1f4001386ce7
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:9e57a66212ac4fca6dd1c802c80cea7e3157157b16577d549f2b1f4001386ce7?se=2022-10-14T08%3A55%3A00Z&sig=zzsKNA3IGyIoFHrwkOUwO3YjOZIQ3%2BSqFG6HhilyOo4%3D&sp=r&spr=https&sr=b&sv=2019-12-12
######################################################################## 100.0%
==> Reinstalling openjpeg 
==> Pouring openjpeg--2.5.0.x86_64_linux.bottle.1.tar.gz
🍺  /home/linuxbrew/.linuxbrew/Cellar/openjpeg/2.5.0: 527 files, 14.1MB

$ brew deps openjpeg
jpeg-turbo
libpng
libtiff
little-cms2
zlib
```

**Reply by kruegercharles** [reply]
Thanks, that helped!
